### ISSUE TITLE
Controller name returned on macOS does not match other platforms

### ISSUE THREAD

**Issue by plink-plonk-will** [godotengine/godot#96756]
### Tested versions

- Reproducible in 4.3. Introduced with commit [07313a08f41146e30005acfa784bdf005d23750b](https://github.com/godotengine/godot/commit/07313a08f41146e30005acfa784bdf005d23750b) 

### System information

Godot v4.3.stable - macOS 14.6.1 - Vulkan (Forward+) - integrated Apple M1 Max - Apple M1 Max (10 Threads)

### Issue description

With the switch to using GCController, the controller name returned no longer matches other platforms. For instance, "PS5 Controller" becomes "DualSense Wireless Controller". This causes a problem when using this value in GDScript to determine controller icon sets across platforms: checking for an additional value on Mac doesn't seem to be a robust solution.

Unfortunately, there doesn't seem to be a way to associate an IOHIDDevice with GCController: so a robust code solution doesn't appear to be possible. SDL works around this by doing string compares and translating the GCController vendorId to the original ids (eg. 0x0ce6 (PS5 controller) 0x054c (Sony)). What I wonder is - should the same thing be done in Godot, or should the differences be handled in the game code? 

_Some notes_:

This is an entry point to seeing how SDL handles this:

https://github.com/libsdl-org/SDL/blob/6e885d96193a4b0096fe7fed6d4e6c3e5f247283/src/joystick/apple/SDL_mfijoystick.m#L387

Here's a stack overflow detailing how you could tell if a IOHIDDevice would be handled by GCController, but I was unable to get the method mentioned to get a valid vendor or product id. I suspect the underlying API has changed, and this is not part of the exposed interface anyway - so even it worked, I suspect it might be risky from an app store review perspective.

https://stackoverflow.com/questions/33509296/supporting-both-gccontroller-and-iohiddeviceref

 

### Steps to reproduce

Use `var controller_name := Input.get_joy_name(device)`, and compare the results from other platforms with that on macOS.

### Minimal reproduction project (MRP)

It's really just the returned value from `Input.get_joy_name(device)`.

**Comment by AdriaandeJongh** [comment]
I suspect this has to do with Apple being the middleman in controller support, whereas on Windows you have 3rd party drivers that are supposedly consistent. I made a workaround for this in my own input manager:
```
func _deduce_controller_type_from_name(controller_name: String) -> ControllerType:
	if controller_name.contains("XInput") or\
			controller_name.contains("Xbox"):
		return ControllerType.XBOX
	elif controller_name.contains("Sony") or\
			controller_name.contains("Playstation") or\
			controller_name.contains("PS4") or\
			controller_name.contains("PS5") or \
			controller_name.contains("DualSense"):
		return ControllerType.PLAYSTATION
	elif controller_name.contains("Joy-Con"):
		if controller_name.contains("(L)"):
			return ControllerType.SWITCH_JOYCON_LEFT
		else:
			return ControllerType.SWITCH_JOYCON_RIGHT
	elif controller_name.contains("Switch"):
		return ControllerType.SWITCH
	elif SteamManager.is_on_steam_deck or controller_name.contains("Steam"):
		return ControllerType.STEAM_DECK
	
	return ControllerType.GENERIC
```

**Comment by plink-plonk-will** [comment]
Yes - that's the workable approach at the moment. The issue is more a question of if that should be done in the engine code instead (like in SDL2). I'm happy to open a PR to do it at the engine level if that's deemed to be the correct approach, but also totally understand if it should be considered non-portable.

**Comment by AdriaandeJongh** [comment]
The workaround I posted above somehow no longer works?! The names I'm getting now are "Controller" and "Wireless Controller", making it impossible to determine the type of controller! Uggghhh. Will make a new issue.

**Comment by plink-plonk-will** [comment]
Which version macOS and Godot are you using? The method used in the link above (https://github.com/rsubtil/controller_icons/issues/83) works for me with macOS 15.0.1, and Godot 4.3 stable.

**Comment by AdriaandeJongh** [comment]
That's the weird thing: I am using Godot v4.3.stable on macOS 15.0.1.

**Comment by Nintorch** [comment]
Hello! May I ask if the issue is still reproducible in Godot 4.5?
